### PR TITLE
ESP Matter + Arduino as IDF Component Light example

### PR DESCRIPTION
**Pull request by SuGlider**
## Description of Change

Adds a new Arduino as IDF component example.
This is a Matter Light example that mixes IDF and Arduino code in order to build an application for a regular Matter Light.

Commissioning can be done using any Matter Provider mobile APP, such as Goggle, Apple, Alexa or ESP Home.
Once it is added to the Matter Fabric, it will can be controlled within the mobile APP.

## Tests scenarios

Tested using Alexa APP + Alexa Dot device and ESP32-S3 as target.
It uses WiFi as Matter transport layer.

## Related links

Follow https://github.com/espressif/arduino-esp32/issues/7432

### CODE REVIEW

**Comment by P-R-O-C-H-Y** [comment]
@SuGlider Is the PR ready to be reviewed? :)

**Comment by SuGlider** [comment]
> @SuGlider Is the PR ready to be reviewed? :)

Not yet.

**Comment by SuGlider** [comment]
@me-no-dev - This shall work fine. Please try it with the Apple Pods and Apple Home APP. Let me know the results.

I have tested it with Alexa Echo 5th Gen.
S3, C3 and C6 work fine with WiFi. 
C6 works fine with Thread, but I had to configure the ESP32 Thread Border Router in order to get Alexa to drive Matter+Thread.

Please task a look into the README.md to see further instructions.

**Review by P-R-O-C-H-Y — APPROVED** [review]
LGTM, tested with ESP32-C3 + Google Home (Nest mini 2nd gen).
As discussed, it would be good to add a note what is needed for Google Home (Developer Console -> new project and what VID/PID to set)

**Review by me-no-dev — APPROVED** [review]

File: idf_component_examples/esp_matter_light/main/idf_component.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+dependencies:
+  espressif/esp_matter:
+    version: "^1.3.0"
+  # Adds Arduino Core from GitHub repository using main branch
+  espressif/arduino-esp32:
+    # version: "^3.0.5"
+    # override_path: "../../../"
+    # pre_release: true
```
Comment (me-no-dev) [inline idf_component_examples/esp_matter_light/main/idf_component.yml:8]: @SuGlider you can remove the comment before merging and releasing 3.0.5

**Comment by SuGlider** [comment]
> LGTM, tested with ESP32-C3 + Google Home (Nest mini 2nd gen). As discussed, it would be good to add a note what is needed for Google Home (Developer Console -> new project and what VID/PID to set)

Done in https://github.com/espressif/arduino-esp32/pull/10290/commits/fe502970b70234f6ef35a7fb4b14be8cb2dfe83d

File: idf_component_examples/esp_matter_light/main/idf_component.yml (hunk 1 above)
Comment (SuGlider) [reply idf_component_examples/esp_matter_light/main/idf_component.yml]: Done. It is ready.

**Comment by me-no-dev** [comment]
@SuGlider don't forget the note to have the iOS device on the same 2.4GHz network

**Comment by SuGlider** [comment]
> @SuGlider don't forget the note to have the iOS device on the same 2.4GHz network

Already done in

b392e8ad6855574890e3f18fd167c852e38bb31b

File: idf_component_examples/esp_matter_light/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+## Returning to the Factory State
+
+Holding the BOOT button pressed for more than 10 seconds and then releasing it will erase all Matter and WiFi configuration, forcing it to reset to factory state. After that, the device needs to be commissioned again. Previous setups done in the Smartphone APP won't work again; therefore, the virtual device shall be removed from the APP.
+
+## Building the Application using WiFi and Matter
+
+Use ESP-IDF 5.1.4 from https://github.com/espressif/esp-idf/tree/release/v5.1  
+This example has been tested with Arduino Core 3.0.4
+
+The project will download all necessary components, inluding the Arduino Core.
```
Comment (SuGlider) [inline idf_component_examples/esp_matter_light/README.md:59]:
```suggestion
The project will download all necessary components, including the Arduino Core.  
```

File: idf_component_examples/esp_matter_light/sdkconfig.defaults.c6_thread [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+# LwIP config for OpenThread
+CONFIG_LWIP_IPV6_NUM_ADDRESSES=8
+CONFIG_LWIP_MULTICAST_PING=y
+
+# MDNS platform
+CONFIG_USE_MINIMAL_MDNS=n
+CONFIG_ENABLE_EXTENDED_DISCOVERY=y
+
+# Enable OTA Requestor
```
Comment (SuGlider) [inline idf_component_examples/esp_matter_light/sdkconfig.defaults.c6_thread:59]:
```suggestion
# Enable OTA Requester
```

File: idf_component_examples/esp_matter_light/sdkconfig.defaults.esp32c3 [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+# Use a custom partition table
+CONFIG_PARTITION_TABLE_CUSTOM=y
+CONFIG_PARTITION_TABLE_FILENAME="partitions.csv"
+CONFIG_PARTITION_TABLE_OFFSET=0xC000
+
+# Disable chip shell
+CONFIG_ENABLE_CHIP_SHELL=n
+
+# Enable OTA Requestor
```
Comment (SuGlider) [inline idf_component_examples/esp_matter_light/sdkconfig.defaults.esp32c3:41]:
```suggestion
# Enable OTA Requester
```

**Review by SuGlider — COMMENTED** [review]
Codespell and Vale fixes.

File: idf_component_examples/esp_matter_light/main/matter_light.cpp [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+  /* Mark deferred persistence for some attributes that might be changed rapidly */
+  cluster_t *level_control_cluster = cluster::get(endpoint, LevelControl::Id);
+  attribute_t *current_level_attribute = attribute::get(level_control_cluster, LevelControl::Attributes::CurrentLevel::Id);
+  attribute::set_deferred_persistence(current_level_attribute);
+
+  cluster_t *color_control_cluster = cluster::get(endpoint, ColorControl::Id);
+  attribute_t *current_x_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentX::Id);
+  attribute::set_deferred_persistence(current_x_attribute);
+  attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id);
```
Comment (SuGlider) [inline idf_component_examples/esp_matter_light/main/matter_light.cpp:267]:
```suggestion
  attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id); // spell-check-ignore
```

**Review by SuGlider — COMMENTED** [review]
Code Spell Check Error `CurrentY --> Currently` solved with
attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id);  // codespell:ignore

File: idf_component_examples/esp_matter_light/main/matter_light.cpp [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+  /* Mark deferred persistence for some attributes that might be changed rapidly */
+  cluster_t *level_control_cluster = cluster::get(endpoint, LevelControl::Id);
+  attribute_t *current_level_attribute = attribute::get(level_control_cluster, LevelControl::Attributes::CurrentLevel::Id);
+  attribute::set_deferred_persistence(current_level_attribute);
+
+  cluster_t *color_control_cluster = cluster::get(endpoint, ColorControl::Id);
+  attribute_t *current_x_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentX::Id);
+  attribute::set_deferred_persistence(current_x_attribute);
+  attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id);  // spell-check-ignore
```
Comment (SuGlider) [inline idf_component_examples/esp_matter_light/main/matter_light.cpp:267]:
```suggestion
  // spell-check-ignore
  attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id);
```

File: idf_component_examples/esp_matter_light/main/matter_light.cpp [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+
+  /* Mark deferred persistence for some attributes that might be changed rapidly */
+  cluster_t *level_control_cluster = cluster::get(endpoint, LevelControl::Id);
+  attribute_t *current_level_attribute = attribute::get(level_control_cluster, LevelControl::Attributes::CurrentLevel::Id);
+  attribute::set_deferred_persistence(current_level_attribute);
+
+  cluster_t *color_control_cluster = cluster::get(endpoint, ColorControl::Id);
+  attribute_t *current_x_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentX::Id);
+  attribute::set_deferred_persistence(current_x_attribute);
+  // spell-check-ignore
```
Comment (P-R-O-C-H-Y) [inline idf_component_examples/esp_matter_light/main/matter_light.cpp:267]:
```suggestion
  attribute::set_deferred_persistence(current_x_attribute); // spell-check-ignore
```

Comment (SuGlider) [inline idf_component_examples/esp_matter_light/main/matter_light.cpp:267]:
```suggestion
  // Ignore Spelling: CurrentY
```

File: idf_component_examples/esp_matter_light/main/matter_light.cpp [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+  /* Mark deferred persistence for some attributes that might be changed rapidly */
+  cluster_t *level_control_cluster = cluster::get(endpoint, LevelControl::Id);
+  attribute_t *current_level_attribute = attribute::get(level_control_cluster, LevelControl::Attributes::CurrentLevel::Id);
+  attribute::set_deferred_persistence(current_level_attribute);
+
+  cluster_t *color_control_cluster = cluster::get(endpoint, ColorControl::Id);
+  attribute_t *current_x_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentX::Id);
+  attribute::set_deferred_persistence(current_x_attribute);
+  // Ignore Spelling: CurrentY
+  attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id);
```
Comment (SuGlider) [inline idf_component_examples/esp_matter_light/main/matter_light.cpp:268]:
```suggestion
   attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id); // codespell:ignore
```

File: idf_component_examples/esp_matter_light/main/matter_light.cpp [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+
+  /* Mark deferred persistence for some attributes that might be changed rapidly */
+  cluster_t *level_control_cluster = cluster::get(endpoint, LevelControl::Id);
+  attribute_t *current_level_attribute = attribute::get(level_control_cluster, LevelControl::Attributes::CurrentLevel::Id);
+  attribute::set_deferred_persistence(current_level_attribute);
+
+  cluster_t *color_control_cluster = cluster::get(endpoint, ColorControl::Id);
+  attribute_t *current_x_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentX::Id);
+  attribute::set_deferred_persistence(current_x_attribute);
+  attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id);  // codespell:ignore
```
Comment (SuGlider) [inline idf_component_examples/esp_matter_light/main/matter_light.cpp:267]:
```suggestion
  attribute_t *current_y_attribute = attribute::get(color_control_cluster, ColorControl::Attributes::CurrentY::Id);  // codespell:ignore
```

**Pull request merged**